### ISSUE TITLE
GEOSConstrainedDelaunayTriangulation_r loops forever on a valid polygon

### ISSUE THREAD

**Issue by nraynaud** [JuliaGeo/LibGEOS.jl#123]
Hi, I reported this issue on GEOS (https://github.com/libgeos/geos/issues/653 ), but it actually seems to be linked to julia.
I have a polygon that sends GEOSConstrainedDelaunayTriangulation_r into a tailspin when called from julia (infinite loop that globles more and more memory over time).

I packaged my test here: https://gist.github.com/nraynaud/6dfbb78d94e1598d0f29d54f70722288

my computer is an Intel macbook pro 
```
$ uname -a
Darwin macreuse.local 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64 x86_64
```

here is my test program: 
```julia 
import LibGEOS

wkt = open("weird_polygon.wkt") do file
    read(file, String)
end
poly = LibGEOS.readgeom(wkt)
@show "before tri"
@show LibGEOS.isValid(poly)
LibGEOS.GEOSConstrainedDelaunayTriangulation_r(LibGEOS._context.ptr, poly.ptr)
@show "after tri"
```

here is the devilish polygon:
![tri](https://user-images.githubusercontent.com/46941/179949859-5843c350-0495-4965-8c34-7b3e987c26ea.png)

```
POLYGON ((1.7862685046962532 0.2923021911868067, 1.7728389809573666 0.2627438719792648, 1.746100000000001 0.2334000000000009, 1.7150804631420686 0.2161586105737922, 1.6759 0.209, 0.3241 0.209, 0.2840389403162291 0.2165244534091589, 0.26281199593527 0.2271117271007482, 0.2451970039279606 0.2409124780256404, 0.220203873371505 0.2753019366857534, 0.2119959512407714 0.2984989878101932, 0.209 0.3233, 0.209 1.6759, 0.2114712418577923 1.6990146526776198, 0.2178177033571282 1.7193153375833403, 0.2407676132817306 1.7547086448098228, 0.2571000000000007 1.7690000000000003, 0.2775992176450858 1.7808316429453195, 0.3030606643061082 1.7889360141633408, 0.324 1.791, 1.676 1.791, 1.715794594594594 1.7835675675675677, 1.7371880040647283 1.7728882728992528, 1.755414722418766 1.7584861946606423, 1.7697324306821693 1.7419033500911634, 1.7821822966428729 1.7193153375833379, 1.7885287581422071 1.699014652677622, 1.7908999998269444 1.6775046880767683, 1.7909999998269444 0.3230046880767684, 1.7862685046962532 0.2923021911868067), (0.691 0.9229754642711925, 0.691 0.7011541640387603, 0.6936627376889104 0.6937321576162595, 0.7020245357288075 0.691, 0.9238458359612397 0.691, 0.9312678423837405 0.6936627376889104, 0.934 0.7020245357288075, 0.934 0.9238458359612397, 0.9313372623110896 0.9312678423837405, 0.9229893209534975 0.9339991155309167, 0.6990442469761068 0.9339012801993407, 0.693329984148253 0.9309041524224672, 0.691 0.9229754642711925), (0.6953392867391299 1.067444033745615, 0.9259557530238928 1.0660987198006593, 0.9315698728030161 1.0689552959301913, 0.9339991155309167 1.0770106790465024, 0.9339012801993407 1.3009557530238929, 0.9315721240202081 1.3060497565109876, 0.9229893209534982 1.3089991155309166, 0.6990442469761072 1.3089012801993407, 0.6934301271969839 1.3060447040698087, 0.691 1.2983866353711833, 0.691 1.0758139687976573, 0.6953392867391299 1.067444033745615), (1.3031380819680736 0.6915341696737235, 1.3089993838262433 0.7013965510439816, 1.3089012068069301 0.9261237482521002, 1.30633726231109 0.93126784238374, 1.2986034489560137 0.9339993838262434, 1.0738762517478988 0.9339012068069299, 1.0687321576162598 0.9313372623110899, 1.0659999999999998 0.923589585046487, 1.0659999999999998 0.7011541640387645, 1.0689552959301913 0.6934301271969839, 1.0770245357288075 0.691, 1.3031380819680736 0.6915341696737235), (1.0690958475775336 1.068329984148252, 1.076410414953513 1.0659999999999998, 1.2991860312023409 1.0659999999999998, 1.3065698728030166 1.0689552959301918, 1.3090000000000002 1.0770245357288104, 1.3090000000000002 1.2988458359612356, 1.3063372623110892 1.306267842383741, 1.2986034489560183 1.3089993838262433, 1.0738762517478997 1.3089012068069301, 1.068732157616259 1.3063372623110892, 1.0660006161737567 1.2986034489560183, 1.0660987931930699 1.0738762517478997, 1.0690958475775336 1.068329984148252))
```

here is the console result on the Apple computer:
```
$ julia --project=. -L test.jl 
"before tri" = "before tri"
LibGEOS.isValid(poly) = true
^C
signal (15): Terminated: 15
in expression starting at /Users/nraynaud/dev/test_delaunay/test.jl:9
_ZN4geos11triangulate7polygon17PolygonHoleJoiner18getLeftShellVertexERKNS_4geom10CoordinateE at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon17PolygonHoleJoiner8joinHoleEPKNS_4geom10LinearRingE at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon17PolygonHoleJoiner9joinHolesEv at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon17PolygonHoleJoiner7computeEv at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon17PolygonHoleJoiner4joinEPKNS_4geom7PolygonE at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon31ConstrainedDelaunayTriangulator18triangulatePolygonEPKNS_4geom7PolygonERNS0_3tri7TriListE at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon31ConstrainedDelaunayTriangulator7computeEv at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
_ZN4geos11triangulate7polygon31ConstrainedDelaunayTriangulator11triangulateEPKNS_4geom8GeometryE at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos.3.10.2.dylib (unknown line)
GEOSConstrainedDelaunayTriangulation_r at /Users/nraynaud/.julia/artifacts/a0d7cbd6dcc9500ac170de3fc30cb2373349205e/lib/libgeos_c.1.16.0.dylib (unknown line)
GEOSConstrainedDelaunayTriangulation_r at /Users/nraynaud/.julia/packages/LibGEOS/VT0Ub/src/libgeos_api.jl:866
unknown function (ip: 0x10a7bd6a6)
jl_apply_generic at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
do_call at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
eval_body at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_interpret_toplevel_thunk at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_toplevel_eval_flex at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_toplevel_eval_flex at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_toplevel_eval_in at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
eval at ./boot.jl:373 [inlined]
include_string at ./loading.jl:1196
jl_apply_generic at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
_include at ./loading.jl:1253
include at ./Base.jl:418 [inlined]
exec_options at ./client.jl:275
_start at ./client.jl:495
jfptr__start_38713.clone_1 at /usr/local/Cellar/julia/1.7.3/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
true_main at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_repl_entrypoint at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
unknown function (ip: 0x0)
Allocations: 1424748 (Pool: 1424062; Big: 686); GC: 2
fatal: error thrown and no exception handler available.
InterruptException()
jl_get_binding_ at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_get_global at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_atexit_hook at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_exit at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)
jl_exit_thread0_cb at /usr/local/Cellar/julia/1.7.3/lib/julia/libjulia-internal.1.7.dylib (unknown line)

```

here is me running the same operation in geos (Apple):
```
$ geosop 
geosop - GEOS 3.11.0
Executes GEOS geometry operations
Usage:
[...]
$ time geosop -a weird_polygon.wkt constrainedDelaunay -f wkt >/dev/null

real	0m0.005s
user	0m0.003s
sys	0m0.001s
```

here is the result in a Linux VM (same underlying computer):
```
$ ~/julia-1.7.3/bin/julia --project=.  -L test.jl 
"before tri" = "before tri"
LibGEOS.isValid(poly) = true
ERROR: LoadError: GEOSError
	std::bad_alloc
Stacktrace:
 [1] geosjl_errorhandler(message::Ptr{UInt8}, userdata::Ptr{Nothing})
   @ LibGEOS ~/.julia/packages/LibGEOS/VT0Ub/src/LibGEOS.jl:106
 [2] GEOSConstrainedDelaunayTriangulation_r(handle::Ptr{Nothing}, g::Ptr{Nothing})
   @ LibGEOS ~/.julia/packages/LibGEOS/VT0Ub/src/libgeos_api.jl:866
 [3] top-level scope
   @ /vagrant/test.jl:9
in expression starting at /vagrant/test.jl:9
vagrant@vagrant:/vagrant$ uname -a
Linux vagrant 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

an extract of Manifest.toml:
```
julia_version = "1.7.3"
manifest_format = "2.0"


[[deps.GEOS_jll]]
deps = ["Artifacts", "JLLWrappers", "Libdl", "Pkg"]
git-tree-sha1 = "4ceb4cdae127931b852ced4d3782bb51ab5e2632"
uuid = "d604d12d-fa86-5845-992e-78dc15976526"
version = "3.10.2+0"


[[deps.LibGEOS]]
deps = ["CEnum", "GEOS_jll", "GeoInterface"]
git-tree-sha1 = "b1506853804fb55b998eb61b66d3afb089dbfa28"
uuid = "a90b1aa1-3769-5649-ba7e-abc5a9d163eb"
version = "0.6.12"

```

Thank you for your nice library.

**Comment by nraynaud** [comment]
I have reviewed a lot of the memory management code in LibGEOS and I want to congratulate the authors on the work you did.

I only found a few minor risks that never materialized in my testing.

**Comment by nraynaud** [comment]
it took me a while to get a compiled and loaded JLL, but here are my results:
```julia
import LibGEOS

@show LibGEOS.GEOSversion()
wkt = open("known_574_bug.wkt") do file
    read(file, String)
end
poly = LibGEOS.readgeom(wkt)
@show "before tri"
@show LibGEOS.isValid(poly)
LibGEOS.GEOSConstrainedDelaunayTriangulation_r(LibGEOS._context.ptr, poly.ptr)
@show "after tri"
```

with the 3.11 JLL loaded:
```
$ julia --project=. test.jl 
LibGEOS.GEOSversion() = "3.11.0-CAPI-1.17.0"
"before tri" = "before tri"
LibGEOS.isValid(poly) = true
"after tri" = "after tri"
```
with the 3.10 JLL loaded:
```
julia> include("../test_delaunay/test.jl")
LibGEOS.GEOSversion() = "3.10.2-CAPI-1.16.0"
"before tri" = "before tri"
LibGEOS.isValid(poly) = true

signal (15): Terminated: 15
```
so it seems to be https://github.com/libgeos/geos/issues/574

what is the schedule process to upgrade the GEOS version underlying this library? 
I don't understand exactly where to send the PR for GEOS_JLL. moreover its version is pinned in LibGEOS.jl

**Comment by nraynaud** [comment]
tried something here: https://github.com/JuliaPackaging/Yggdrasil/pull/5203

**Comment by visr** [comment]
Thanks for digging into the issue, tracking it to a fixed GEOS issue, and submitting the Yggdrasil PR.

> what is the schedule process to upgrade the GEOS version underlying this library? I don't understand exactly where to send the PR for GEOS_JLL. moreover its version is pinned in LibGEOS.jl

GEOS 3.11 is about 3 weeks old now. There is no schedule upgrade process, it was basically waiting for somebody to make the Yggdrasil PR, so thanks for doing that. Note that `GEOS_jll = "3.10"` does not mean it is pinned, it is simply 3.10 or higher (but not 4). So once the new JLL is out, it would be nice to re-run the wrapper generator code under `gen/` such that the new functions from 3.11 get wrapped. When we do this we should also upgrade the minimum required JLL version to 3.11, to make sure that those new functions actually exist in the provided JLL.

**Comment by visr** [comment]
Now that GEOS_jll 3.11 is available in the registry, I ran the tests of this package to make sure all is well. All tests are still passing, so that is great. My point about re-running the wrapper generator is automatically suggested when running the tests:

```
┌ Warning: The GEOS version in use is different from the version used to generate libgeos_api.jl. Consider regenerating if the difference is more than a patch release.
│   run_version = "3.11.0-CAPI-1.17.0"
│   gen_version = "3.10.2-CAPI-1.16.0"
```

Note that GDAL_jll is does pin GEOS_jll and PROJ_jll to the minor version. This is still 3.10. I'm not sure yet if they would be compatible or not. But in any case it'd be good to update the GDAL build to use the latest PROJ and GEOS, such that GEOS 3.11 can work together with GDAL in the same project.

https://github.com/JuliaBinaryWrappers/GDAL_jll.jl/blob/ab9147cf879908240f4b0a946d4901a91664bb1d/Project.toml#L24-L27

**Comment by nraynaud** [comment]
thanks, I'll try to run the generator and send a PR here first. 
I am a bit dazzled, I have now 2 full months of professional experience in julia under my belt, and tracking what I thought was a memory issue at the GC/malloc interface felt a bit early.

**Comment by visr** [comment]
Fixed by #124
